File: leetcode/typescript/000704.binary-search.ts
function search(nums: number[], target: number): number {
	let l = 0;
	let r = nums.length - 1;

	while (l <= r) {
		const mid = Math.floor((r + l) / 2);
		if (nums[mid] < target) {
			l = mid + 1;
		} else if (nums[mid] > target) {
			r = mid - 1;
		} else {
			return mid;
		}
	}

	return -1;
}
// Time complexity: O(n)
// Space complexity: O(1)

File: leetcode/python/000704.binary-search.py
class Solution:
    # Time complexity: O(log(n))
    # Space complexity: O(1)
    def search(self, nums: list[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l < r:
            m = (r - l) // 2

            if nums[m] < target:
                l = m + 1
            elif nums[m] > target:
                r = m - 1
            else:
                return m

        return -1
